File: GOF/Adapter/cpp/TextShapeObjectAdapter.cpp
//
// Created by 17271 on 2019/3/27.
//

#include "TextShapeObjectAdapter.h"
#include "Coord.h"
#include "TextView.h"
#include "Point.h"
#include "TextManipulator.h"

using std::make_shared;

TextShapeObjectAdapter::TextShapeObjectAdapter(std::shared_ptr<TextView> textView)
	: m_textView{textView}
{}

void TextShapeObjectAdapter::boundingBox(Point &bottomLeft, Point &topRight) const
{
	Coord bottom, left, width, height;

	m_textView->getOrigin(bottom, left);
	m_textView->getExtent(width, height);

	bottomLeft = Point(bottom, left);
	topRight = Point(bottom + height, left + width);
}

std::shared_ptr<Manipulator> TextShapeObjectAdapter::createManipulator() const
{
	return make_shared<TextManipulator>();
}

bool TextShapeObjectAdapter::isEmpty() const
{
	return m_textView->isEmpty();
}

File: GOF/Adapter/java/src/TextShapeObjectAdapter.java
public class TextShapeObjectAdapter extends Shape {

	private TextView mTextView;

	public TextShapeObjectAdapter(TextView textView) {
		mTextView = textView;
	}

	@Override
	public void boundingBox(Point bottomLeft, Point topRight) {
		Coord bottom = new Coord();
		Coord left = new Coord();
		Coord width = new Coord();
		Coord height = new Coord();

		mTextView.getOrigin(bottom, left);
		mTextView.getExtent(width, height);

		bottomLeft.set(bottom, left);
		topRight.set(bottom.add(height), left.add(width));
	}

	@Override
	public Manipulator createManipulator() {
		return new TextManipulator();
	}

	public boolean isEmpty() {
		return mTextView.isEmpty();
	}
}
